### DISCUSSION TITLE
Trouble Flipping Magnetization of Object

### DISCUSSION THREAD

**Discussion by leahkong02** [magpylib/magpylib#570] [Q&A] [answered]
I'm trying to simulate magnetic field of six axially magnetized ring magnets, where there are 3 pairs of 2 oppositely magnetized rings in each of the three coordinate directions. However, when I'm doing both the graphical output and magnetic field calculations, the field and magnetization in the y direction does not seem correct. The y-direction keeps behaving like two same polarized magnetized rings instead of opposite polarity. I'm not sure if this is a result of my using rotation objects incorrectly, as I haven't used scipy rotation objects before this, but my rotation seems to work in the x direction, and I've tried to change the rotation object to different permutations of 90 degrees about the x axis. Here is my code: 

```python
import magpylib as magpy
from scipy.spatial.transform import Rotation as R

inner_r = 60
outer_r = 35
h = 10
mag = [0, 0, 1000]
nmag = [0, 0, -1000]
distance = 70
ndistance = -70
dim = (outer_r, inner_r, h, 0, 360)

a = magpy.magnet.CylinderSegment(mag, dim, [0, 0, distance])
b = magpy.magnet.CylinderSegment(nmag, dim, [0, 0, ndistance])
c = magpy.magnet.CylinderSegment(mag, dim, [0, distance, 0])
d = magpy.magnet.CylinderSegment(nmag, dim, [0, ndistance, 0])
e = magpy.magnet.CylinderSegment(mag, dim, [distance, 0, 0])
f = magpy.magnet.CylinderSegment(nmag, dim, [ndistance, 0, 0])

c.orientation = R.from_euler('x', 90, degrees = True)
d.orientation = R.from_euler('x', 90, degrees = True)
e.orientation = R.from_euler('y', 90, degrees = True)
f.orientation = R.from_euler('y', 90, degrees = True)

rings = magpy.Collection(a, b, c, d, e, f)
magpy.show(rings)
```
And here is what the graphics look like:
![Figure_1](https://user-images.githubusercontent.com/74679331/211728586-9f81daee-4bce-4abd-a10b-f2e6876a34cf.png)

Any help navigating this issue would be much appreciated!

**Comment by OrtnerMichael** [comment, marked as answer]
hi @leahkong02 

Thanks for reaching out. There are two problems
1. There is a Bug in our Matplotlib graphic output
2. There is also a Bug in your code (the first rotation should be with respect to -90°)

Both together lead to confusion that is however easily solved !
Your correct code should be like this:

``` python
import magpylib as magpy

inner_r = 60
outer_r = 35
h = 10
mag = [0, 0, 1000]
nmag = [0, 0,-1000]
distance = 70
dim = (outer_r, inner_r, h, 0, 360)

a = magpy.magnet.CylinderSegment( mag, dim, [0, 0,  distance])
b = magpy.magnet.CylinderSegment(nmag, dim, [0, 0, -distance])
c = magpy.magnet.CylinderSegment( mag, dim, [0,  distance, 0])
d = magpy.magnet.CylinderSegment(nmag, dim, [0, -distance, 0])
e = magpy.magnet.CylinderSegment( mag, dim, [ distance, 0, 0])
f = magpy.magnet.CylinderSegment(nmag, dim, [-distance, 0, 0])

c.rotate_from_angax(-90, 'x')
d.rotate_from_angax(-90, 'x')
e.rotate_from_angax(90, 'y')
f.rotate_from_angax(90, 'y')

for x in [a,b,c,d,e,f]:
    print(x.getB((0,0,0)))

magpy.show(a,b,c,d,e,f, backend='plotly')
```
The output should be

[ 0.          0.         10.17666237]
[  0.           0.         -10.17666237]      
[ 0.         10.17666237  0.        ]
[  0.         -10.17666237   0.        ]      
[10.17666237  0.          0.        ]
[-10.17666237   0.           0.        ] 

which shows that the field computation is good, and the graphic


![image](https://user-images.githubusercontent.com/47384012/211814842-1df6b7c1-cd19-42eb-84bf-7893a6bcb302.png)

Here the graphic output is created with the optional Plotly packge. Matplotlib will still give a bad output - this will be fixed with the next update.

does this answer your question ?

**Reply by leahkong02** [reply]
this helps, thank you so much!! magpylib has been super useful in my research, and I really appreciate your prompt response :)

**Reply by OrtnerMichael** [reply]
glad to help :)

**Comment by Alexboiboi** [comment]
Hi @leahkong02, thanks for your feedback about the library ;)

-  The display issue is tracked here #571 and will be fixed in the coming release, very soon.

**Comment by Alexboiboi** [comment]
Hi @leahkong02, we have released a patch version that also fixes the wrong display of magnetization arrows like in this case.

you can upgrade your version with 
```bash
pip install -U magpylib
```
or by giving the explicit version number

```bash
pip install magpylib==4.1.2
```

or if you use conda:

```bash
conda install -c conda-forge magpylib=4.1.2
```

:-)

**Reply by leahkong02** [reply]
Thank you so much!!
